Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -294,44 +258,96 @@ internal final class AsyncServerHandler<
 
   @inlinable
   internal func receiveMetadata(_ headers: HPACKHeaders) {
-    self.interceptors.receive(.metadata(headers))
+    switch self.interceptorStateMachine.interceptRequestMetadata() {
+    case .intercept:
+      self.interceptors?.receive(.metadata(headers))
+    case .cancel:
+      self.cancel(error: nil)
+    case .drop:
+      ()
+    }
   }
 
   @inlinable
   internal func receiveMessage(_ bytes: ByteBuffer) {
+    let request: Request
+
     do {
-      let message = try self.deserializer.deserialize(byteBuffer: bytes)
-      self.interceptors.receive(.message(message))
+      request = try self.deserializer.deserialize(byteBuffer: bytes)
     } catch {
-      self.handleError(error)
+      return self.cancel(error: error)
+    }
+
+    switch self.interceptorStateMachine.interceptRequestMessage() {
+    case .intercept:
+      self.interceptors?.receive(.message(request))
+    case .cancel:
+      self.cancel(error: nil)
+    case .drop:
+      ()
     }
   }
 
   @inlinable
   internal func receiveEnd() {
-    self.interceptors.receive(.end)
+    switch self.interceptorStateMachine.interceptRequestEnd() {
+    case .intercept:
+      self.interceptors?.receive(.end)
+    case .cancel:
+      self.cancel(error: nil)
+    case .drop:
+      ()
+    }
   }
 
   @inlinable
   internal func receiveError(_ error: Error) {
-    self.handleError(error)
-    self.finish()
+    self.cancel(error: error)
   }
 
   @inlinable
   internal func finish() {
-    switch self.state {
-    case .idle:
-      self.interceptors = nil
-      self.state = .completed
+    self.cancel(error: nil)
+  }
 
-    case .active:
-      self.state = .completed
-      self.interceptors = nil
-      self.userHandlerTask?.cancel()
+  @usableFromInline
+  internal func cancel(error: Error?) {
+    self.eventLoop.assertInEventLoop()
+
+    switch self.handlerStateMachine.cancel() {
+    case .cancelAndNilOutHandlerComponents:
+      // Cancel handler related things (task, response writer).
+      self.handlerComponents?.cancel()
+      self.handlerComponents = nil
+
+      // We don't distinguish between having sent the status or not; we just tell the interceptor
+      // state machine that we want to cancel. It will inform us whether to generate and send a
+      // status or not.
+      switch self.interceptorStateMachine.interceptResponseStatus() {
```

Comment:
I ummed and ahhed about this. The rationale is that if we're mid-RPC then we should let the interceptors know the rpc is going away. The downside is that it requires the interceptors to play ball and not do anything async. Clearly I didn't do a good job of convincing myself because on L344 we just send out the status if we cancel.